flink-end-to-end-tests/flink-stream-state-ttl-test/src/main/java/org/apache/flink/streaming/tests/MonotonicTTLTimeProvider.java

```diff
@@ -54,14 +57,24 @@
 	private static final Object lock = new Object();
 
 	@GuardedBy("lock")
-	static long freeze() {
+	static <T, E extends Throwable> T doWithFrozenTime(FunctionWithException<Long, T, E> action) throws E {
 		synchronized (lock) {
-			if (!timeIsFrozen || lastReturnedProcessingTime == Long.MIN_VALUE) {
-				timeIsFrozen = true;
-				return getCurrentTimestamp();
-			} else {
-				return lastReturnedProcessingTime;
-			}
+			final long timestampBeforeUpdate = freeze();
+			T result = action.apply(timestampBeforeUpdate);
```

Comment:
I fixed one more thing here. We have to use the frozen time `timestampBeforeUpdate` because if we use `getCurrentTimestamp` it will change the frozen time and fail `checkState` later.